mlflow/sklearn/__init__.py

```diff
@@ -455,6 +458,146 @@ def load_model(model_uri):
+       flags for method `model.fit`, `eval_and_log_metrics`, `model.score`.
+       In order to avoid nested/duplicated autologging metric, when run into these scopes,
+       we need temporarily disable the metric autologging.
+    """
+
+    fit_scope_key = 'fit'
+    eval_and_log_metrics_scope_key = 'eval_and_log_metrics'
+    model_score_scope_key = 'model_score'
+
+    def __init__(self):
+        self.model_id_to_run_id_map = {}
+        self.pred_result_id_to_dataset_name_and_model_id = {}
+        self.scope_flag_dict = {
+            self.fit_scope_key: False,
+            self.eval_and_log_metrics_scope_key: False,
+            self.model_score_scope_key: False
+        }
+
+    def call_scope(self, scope_key):
+        class CallScope:
+            def __enter__(inner_self):
+                self.scope_flag_dict[scope_key] = True
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self.scope_flag_dict[scope_key] = False
+
+        return CallScope()
+
+    def get_run_id_for_model(self, model):
+        model_uuid = _get_obj_uuid(model)
+        if model_uuid:
+            return self.model_id_to_run_id_map.get(model_uuid, None)
+        else:
+            return None
+
+    def should_log_eval_metrics(self):
+        return not self.scope_flag_dict[self.fit_scope_key] and \
+            not self.scope_flag_dict[self.eval_and_log_metrics_scope_key] and \
+            not self.scope_flag_dict[self.model_score_scope_key]
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        obj, model_uuid = _assign_obj_uuid(model)
+        if obj is model and model_uuid:
+            self.model_id_to_run_id_map[model_uuid] = run_id
+
+    def register_prediction_result(self, model, eval_dataset_name, predict_result):
+        """
+        In `patched_predict`, register the prediction result instance with the model id and eval
+        dataset name. e.g.
+
+        ```
+        prediction_result = model_1.predict(eval_X)
+        ```
+        then we need register an item into the `pred_result_id_to_dataset_name_and_model_id` map like
+        below:
+        id(prediction_result) --> (inspected_original_var_name_of(eval_X), id(model_1))
+
+        With this map `dataset_id_to_dataset_name_and_model_id`, in following patched metric API,
+        we can query the eval dataset name and the model id via the `y_pred` argument instance.
+        """
+        model_uuid = _get_obj_uuid(model)
+        print('DGB: register_prediction_result # 1')
+        if model_uuid:
+            print('DGB: register_prediction_result # 2')
+            value = (eval_dataset_name, model_uuid)
+            predict_result, prediction_result_uuid = _assign_obj_uuid(predict_result)
```

Comment:
`prediction_result_uuid` is no longer an instance of `np.ndarray` but `_NumpyArrayWithUUID`. Could this be an issue?

Reply:
`isinstance(predict_result)` still returns True though

Reply:
_NumpyArrayWithUUID will inherits all nd.array methods. I think it should be fine.
seems no other way to attach uuid.

Reply:
I think it's rare to do this but this breaks when `predict` is a `_NumpyArrayWithUUID` object.

```python
preds = model.predict(...)

if type(preds) == np.ndarray:
    # do something
```

Reply:
Since we know that models can always be assigned attributes and that numpy arrays are the only source of concern (that we know of), can we still maintain an `id(numpy_array)` mapping for numpy arrays?

Here's the scenario that I think your logic is trying to avoid:

1. A user trains `Model1`, which is associated with `Run1`
2. The user evaluates `Model1` on test data `test_x`, producing prediction result `test_y`.
3. `test_y` is garbage collected.
4. Another object is created with the same ID as `test_y`. Let's call it `test_y_impostor`.
5. `test_y_impostor` is passed to `accuracy_score()`, which logs a metric to `Run1`, even though `test_y_impostor` didn't come from `Model1`.

**but**, we need to ask ourselves: "How was `test_y_impostor` generated?". Most likely, `test_y_impostor` came from another call to `Model2.predict()` during a new run such as `Run2`, right? Such a call to `model.predict()` would reassign the ID mapping to `Run2` before we ever get to step (5). Therefore, I don't think this is an issue.

The reason I'd prefer to not subclass `np.ndarray` is that some applications may actually rely on this specific return type, and returning a subclass causes us to degrade the safety of our patching infrastructure.

Let me know if I've missed something here. In summary, I think we should use attribute assignment for models and ID mapping for numpy arrays (or any other type of object that doesn't support attribute assignment).

Reply:
Rethink on this, a key point:
* in `patched_metric_api`, we don't have "model object" context (metric api is not model's method), so that we cannot get the "run_id" directly, we have to query the `pred_result_id_to_dataset_name_and_model_id` map to get the corresponding model id. Note that `pred_result_id_to_dataset_name_and_model_id` use "prediction_result_object_id" as key, if the id being reused, then will leads to wrong mapping.
* Metric api call arguments not only include "prediction_result_object", it also include other arguments (e.g. `y_true` arguments), a new `y_true` object may reuse the ID of an GCed "prediction_result_object". And, in our code, we don't know which metric call argument is the "prediction_result" argument, we try on every argument's ID to query the `pred_result_id_to_dataset_name_and_model_id`. Then suppose the case: in following metric api call, a new `y_true` argument object reuse the ID of an GCed "prediction_result_object", then it break the mapping relationship.

So, addressing the ID reusing issue is still important. Fortunately, I found a way, using `weakref.finalize` like:
```
        def clean_id(id_):
            _get_autolog_training_status().pred_result_id_to_dataset_name_and_model_id \
                .pop(id_, None)

        # When the `predict_result` object being GCed, its ID may be reused, so register a finalizer
        # to clear the ID from the dict for preventing wrong ID mapping.
        weakref.finalize(predict_result, clean_id, prediction_result_id)
```
we can address the issue.

Reply:
Provide an example to demonstrate how reusing ID leads to wrong result:

```
model1.fit(...) # suppose create mlflow run_id1
eval_X_1, eval_y_1 = load_data1(...)
pred_y_1 = model1.predict(eval_X_1) # suppose `id(pred_y_1)`=10,
                                                             # register key=10 into `pred_result_id_to_dataset_name_and_model_id`
accuracy_score(eval_y_1, pred_y_1) # will log metric into "run_id1"
...
del pred_y_1 # `pred_y_1` will be GCed, id=10 was released

model2.fit(...) # suppose create mlflow run_id2
eval_X_2, eval_y_2 = load_data2(...) # suppose `eval_y_2` reuse id=10
pred_y_2 = model2.predict(eval_X_2) # suppose `id(pred_y_2)`=20
accuracy_score(eval_y_2, pred_y_2) # Now we will query `pred_result_id_to_dataset_name_and_model_id` by
                                                             # both `id(eval_y_2)`(=10) and `id(pred_y_2)`(=20),  we will match 
                                                             # `id(eval_y_2)` first (it reuse ID=10 which previously was `pred_y_1` 
                                                             # object, and then we query out run_id1, then log metric into run_id1.







```